cpp/src/arrow/record_batch.cc

```diff
@@ -487,6 +487,27 @@ struct EnumeratedStatistics {
 };
 using OnStatistics =
     std::function<Status(const EnumeratedStatistics& enumerated_statistics)>;
+using ArrayStatisticsAndTypeVector =
+    std::vector<std::pair<std::shared_ptr<ArrayStatistics>, std::shared_ptr<DataType>>>;
+ArrayStatisticsAndTypeVector ExtractArrayDataAndType(const RecordBatch& record_batch) {
+  ArrayDataVector traverse;
+  ArrayStatisticsAndTypeVector result;
+  for (int i = 0; i < record_batch.num_columns(); ++i) {
+    auto column_array_data = record_batch.column(i)->data();
+    result.emplace_back(column_array_data->statistics, column_array_data->type);
+    traverse.insert(traverse.end(), column_array_data->child_data.crbegin(),
+                    column_array_data->child_data.crend());
+    while (!traverse.empty()) {
+      auto child_data = traverse.back();
+      traverse.pop_back();
+      result.emplace_back(child_data->statistics, child_data->type);
+      traverse.insert(traverse.end(), child_data->child_data.crbegin(),
+                      child_data->child_data.crend());
+    }
```

Comment:
IPC writer has max recursion check:
https://github.com/apache/arrow/blob/c3e399af43c4a2e384a41fad0619589baa9045f0/cpp/src/arrow/ipc/writer.cc#L149-L151
https://github.com/apache/arrow/blob/c3e399af43c4a2e384a41fad0619589baa9045f0/cpp/src/arrow/ipc/options.h#L49-L50
https://github.com/apache/arrow/blob/c3e399af43c4a2e384a41fad0619589baa9045f0/cpp/src/arrow/ipc/options.h#L37-L40

We may need similar check but how to make it customizable? We may need `MakeStatisticsOptions` or something...